File: data_structure/priority_queue/295_find_median_from_data_stream.cc
#include <queue>

// 2024/01/14: 有思路, 但是实现看了答案.
class MedianFinder {
 public:
  MedianFinder() {}

  void AddNum(int num) {
    if (left_pq.empty() || num <= left_pq.top()) {
      left_pq.push(num);
      if (right_pq.size() + 1 < left_pq.size()) {
        right_pq.push(left_pq.top());
        left_pq.pop();
      }
    } else {
      right_pq.push(num);
      if (right_pq.size() > left_pq.size()) {
        left_pq.push(right_pq.top());
        right_pq.pop();
      }
    }
  }

  double FindMedian() {
    if (left_pq.size() > right_pq.size()) {
      return left_pq.top();
    }
    return (left_pq.top() + right_pq.top()) / 2.0;
  }

 private:
  // 注意两个优先队列的定义.
  std::priority_queue<int, std::vector<int>, std::less<int>> left_pq;
  std::priority_queue<int, std::vector<int>, std::greater<int>> right_pq;
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->AddNum(num);
 * double param_2 = obj->FindMedian();
 */

File: data_structure/priority_queue/295_find_median_from_data_stream.py
import heapq


# 2026/06/03: 自己实现.
class MedianFinder:

    def __init__(self):
        self.left_max_priority_queue = []
        self.right_min_priority_queue = []

    def add_num(self, num: int) -> None:
        if not self.left_max_priority_queue:
            heapq.heappush(self.left_max_priority_queue, -num)
        else:
            if num <= -self.left_max_priority_queue[0]:
                heapq.heappush(self.left_max_priority_queue, -num)
            else:
                heapq.heappush(self.right_min_priority_queue, num)

        if len(self.left_max_priority_queue) < len(self.right_min_priority_queue):
            right_top = heapq.heappop(self.right_min_priority_queue)
            heapq.heappush(self.left_max_priority_queue, -right_top)
        elif len(self.left_max_priority_queue) > len(self.right_min_priority_queue) + 1:
            left_top = -heapq.heappop(self.left_max_priority_queue)
            heapq.heappush(self.right_min_priority_queue, left_top)

    def find_median(self) -> float:
        if len(self.left_max_priority_queue) > len(self.right_min_priority_queue):
            left = -self.left_max_priority_queue[0]
            return left
        else:
            left = -self.left_max_priority_queue[0]
            right = self.right_min_priority_queue[0]
            return (left + right) / 2


obj = MedianFinder()
obj.add_num(1)
obj.add_num(2)
print(obj.find_median())
obj.add_num(3)
print(obj.find_median())
